modules/miner/testminer.go

```diff
@@ -0,0 +1,96 @@
+package miner
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"unsafe"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// testminer.go implements the TestMiner interface, whose primary purpose is
+// integration testing.
+
+// BlockForWork returns a block that is ready for nonce grinding, along with
+// the root hash of the block.
+func (m *Miner) BlockForWork() (b types.Block, merkleRoot crypto.Hash, t types.Target, err error) {
+	// Check if the wallet is unlocked. If the wallet is unlocked, make sure
+	// that the miner has a recent address.
+	if !m.wallet.Unlocked() {
+		err = modules.ErrLockedWallet
+		return
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	err = m.checkAddress()
+	if err != nil {
+		return
+	}
+
+	b, t = m.blockForWork()
+	merkleRoot = b.MerkleRoot()
+	return b, merkleRoot, t, nil
+}
+
+// AddBlock adds a block to the consensus set.
+func (m *Miner) AddBlock() (types.Block, error) {
+	block, err := m.FindBlock()
+	if err != nil {
+		return types.Block{}, err
+	}
+	err = m.cs.AcceptBlock(block)
+	if err != nil {
+		return types.Block{}, err
+	}
+	return block, nil
+}
+
+// FindBlock finds at most one block that extends the current blockchain.
+func (m *Miner) FindBlock() (types.Block, error) {
+	m.mu.Lock()
+	if !m.wallet.Unlocked() {
+		return types.Block{}, modules.ErrLockedWallet
+	}
+	err := m.checkAddress()
+	if err != nil {
+		return types.Block{}, err
+	}
+	m.mu.Unlock()
+
+	// Get a block for work.
+	m.mu.Lock()
+	bfw, target := m.blockForWork()
+	m.mu.Unlock()
+
+	block, ok := m.SolveBlock(bfw, target)
+	if !ok {
+		return types.Block{}, errors.New("could not solve block using limited hashing power")
+	}
+	return block, nil
+}
+
+// SolveBlock takes a block, target, and number of iterations as input and
+// tries to find a block that meets the target. This function can take a long
+// time to complete, and should not be called with a lock.
+func (m *Miner) SolveBlock(b types.Block, target types.Target) (types.Block, bool) {
+	// Assemble the header.
+	merkleRoot := b.MerkleRoot()
+	header := make([]byte, 80)
+	copy(header, b.ParentID[:])
+	binary.LittleEndian.PutUint64(header[40:48], uint64(b.Timestamp))
+	copy(header[48:], merkleRoot[:])
+
+	nonce := (*uint64)(unsafe.Pointer(&header[32]))
+	for i := 0; i < iterationsPerAttempt; i++ {
+		id := crypto.HashBytes(header)
+		if bytes.Compare(target[:], id[:]) >= 0 {
+			copy(b.Nonce[:], header[32:40])
+			return b, true
+		}
+		*nonce++
```

Comment:
this probably makes little difference, but I found that you can speed up this increment nearly 3x in a simple benchmark:
```go
// Benchmark says: 2.40 ns/op
var header [80]byte
nonce := (*uint64)(unsafe.Pointer(&header[32]))
for i := 0; i < b.N; i++ {
    *nonce++ // note that this is equivalent to *nonce = *nonce + 1
}

// Benchmark says: 0.77 ns/op
var header [80]byte
nonce := (*uint64)(unsafe.Pointer(&header[32]))
var uintNonce uint64 // local variable that matches values of nonce
for i := 0; i < b.N; i++ {
    uintNonce++
    *nonce = uintNonce // only one deref instead of two
}
```
For comparison, just incrementing a local uint64 takes 0.41 ns/op.